File: Homework/themoviedb/src/main/java/com/example/themoviedb/SearchActivity.java
package com.example.themoviedb;

import android.annotation.SuppressLint;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.themoviedb.models.MovieResponse;
import com.example.themoviedb.recyclerView.SearchMovieAdapter;
import com.example.themoviedb.repository.RepositoryProvider;

import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;

public class SearchActivity extends AppCompatActivity {

    @BindView(R.id.search_movie_recyclerView)
    RecyclerView recyclerView;

    @BindView(R.id.searchText)
    SearchView searchTextView;

    SearchMovieAdapter adapter;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.search_activity);
        ButterKnife.bind(this);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        adapter = new SearchMovieAdapter();

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                layoutManager.getOrientation());
        recyclerView.addItemDecoration(dividerItemDecoration);

        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);

        getObservableQuery(searchTextView)
                .debounce(2, TimeUnit.SECONDS)
                .filter(text -> !text.isEmpty() && text.length() >= 3)
                .distinctUntilChanged()
                .switchMap((Function<String, ObservableSource<MovieResponse>>) query -> RepositoryProvider.get()
                        .provideRepository()
                        .getMovies(query))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(MovieResponse::getMovies)
                .subscribe(adapter::setMovies);
}

    private Observable<String> getObservableQuery(SearchView searchTextView){
        PublishSubject<String> publishSubject = PublishSubject.create();

        searchTextView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String newText) {
                publishSubject.onNext(newText);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                publishSubject.onNext(newText);
                return true;
            }
        });

        return publishSubject;
    }
}

File: Homework/themoviedb/src/main/java/com/example/themoviedb/SearchActivity.kt
package com.example.themoviedb

import android.annotation.SuppressLint
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import butterknife.ButterKnife
import com.example.themoviedb.recyclerView.SearchMovieAdapter
import com.example.themoviedb.repository.RepositoryProvider
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.search_activity.*
import java.util.concurrent.TimeUnit

@Suppress("UNCHECKED_CAST")
class SearchActivity : AppCompatActivity() {

    @SuppressLint("CheckResult")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.search_activity)
        ButterKnife.bind(this)

        val layoutManager = LinearLayoutManager(this)
        val adapter = SearchMovieAdapter()

        val dividerItemDecoration = DividerItemDecoration(searchMovieRecyclerView.context,
                layoutManager.orientation)
        searchMovieRecyclerView.addItemDecoration(dividerItemDecoration)

        searchMovieRecyclerView.setHasFixedSize(true)
        searchMovieRecyclerView.layoutManager = layoutManager
        searchMovieRecyclerView.adapter = adapter

        getObservableQuery(searchText)
                .debounce(2, TimeUnit.SECONDS)
                .filter { text -> text.isNotEmpty() && text.length >= 3 }
                .distinctUntilChanged()
                .switchMap { query: String ->
                    RepositoryProvider.get()!!
                            .provideRepository()
                            .getMovies(query)
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map { it.movies }
                .subscribe { adapter.setMovies(it) }
    }

    private fun getObservableQuery(searchTextView: SearchView): Observable<String> {
        val publishSubject = PublishSubject.create<String>()

        searchTextView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(newText: String): Boolean {
                publishSubject.onNext(newText)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                publishSubject.onNext(newText)
                return true
            }
        })

        return publishSubject
    }
}